Lib/asyncio/base_events.py

```diff
@@ -795,6 +801,49 @@ def _getaddrinfo_debug(self, host, port, family, type, proto, flags):
 
         return transport, protocol
 
+    async def start_tls(self, transport, protocol, sslcontext, *,
+                        server_side=False,
+                        ssl_handshake_timeout=None):
+        """Upgrade transport to TLS.
+
+        Return a new transport that *protocol* should start using
+        immediately.
+        """
+        if ssl is None:
+            raise RuntimeError('Python ssl module is not available')
+
+        if not isinstance(sslcontext, ssl.SSLContext):
+            raise TypeError(
+                f'sslcontext is expected to be an instance of ssl.SSLContext, '
+                f'got {sslcontext!r}')
+
+        if not getattr(transport, '_start_tls_compatible', False):
+            raise TypeError(
+                f'transport {self!r} is not supported by start_tls()')
+
+        if transport.get_write_buffer_size():
+            if transport._buffer_drained_fut is not None:
```

Comment:
Is it necessary to rely on an implementation detail here?

Reply:
Yes.  Transports are intimately tied to event loops and are aware of each other private details.  Another event loop, say uvloop, can't support asyncio `Transport` classes, and vice versa.

Currently, transports don't have APIs to notify the loop when their write buffers are drained.  We probably need to design such a public API, but that's out of the scope of this PR.

That said, I think I'll add a private base class to share some functionality between Proactor and Selector transport classes. That will also make code a bit prettier.